### PR TITLE
client: add retryable error and refactor client tests

### PR DESCRIPTION
**Pull request by morefreeze**
@siddontang @disksing PTAL

### CODE REVIEW

File: client.go [hunk 1]
```diff
@@ -33,9 +34,17 @@ type Client struct {
...
 }
 
-const initConnection int = 5
-const maxConnecion int = 100
-const maxRetryGetConn int = 5
+const (
+	initConnection  int           = 5
+	maxConnecion    int           = 100
+	maxRetryGetConn int           = 5
+	readTimeout     time.Duration = 5 // seconds
```
Comment (siddontang) [inline client.go]: writeTimeout missing?

File: client_test.go [hunk 2]
```diff
@@ -98,3 +62,55 @@ func (s *testClientSuite) testSendBySelf(c *C) {
...
+}
+
+func startServer(host string, c *C, handleFunc func(net.Conn, *C)) net.Listener {
+	l, err := net.Listen("tcp", host)
+	c.Assert(err, IsNil)
+	log.Debug("Start listenning on", host)
+	go func() {
+		for {
+			conn, err := l.Accept()
+			c.Assert(err, IsNil)
```
Comment (siddontang) [inline client_test.go]: no need to check err, if err, return is ok.

File: client.go [hunk 3]
```diff
@@ -33,9 +34,17 @@ type Client struct {
...
+)
+
+var (
+	// ErrRetryable return upper level to decide weather retry or not.
```
Comment (disksing) [inline client.go]: whether

File: client.go [hunk 4]
```diff
@@ -99,11 +106,13 @@ func (c *Client) Close() error {
...
 	curMsgID := atomic.AddUint64(&c.msgID, 1)
 	log.Debugf("Send request msgID[%d] type[%s]", curMsgID, req.GetType())
-	err := kvrpc.EncodeMessage(conn, curMsgID, req)
-	if err != nil {
+	if err := kvrpc.EncodeMessage(conn, curMsgID, req); err != nil {
 		return nil, errors.Trace(err)
 	}
 	resp := new(pb.Response)
+	if err := conn.SetReadDeadline(time.Now().Add(time.Second * readTimeout)); err != nil {
+		log.Warn("Set write deadline failed, it may be blocked.")
```
Comment (siddontang) [inline client.go]: read deadline

File: client.go [hunk 5]
```diff
@@ -33,9 +34,17 @@ type Client struct {
 	pool  pool.Pool
 }
 
-const initConnection int = 5
-const maxConnecion int = 100
-const maxRetryGetConn int = 5
+const (
+	initConnection  int           = 5
+	maxConnecion    int           = 100
+	maxRetryGetConn int           = 5
```
Comment (disksing) [inline client.go]: Is this an unused constant?

File: client.go (hunk 1 above)
Comment (morefreeze) [inline client.go]: will add it.

File: client_test.go (hunk 2 above)
Comment (morefreeze) [inline client_test.go]: got it.

File: client.go (hunk 5 above)
Comment (morefreeze) [inline client.go]: will remove it.

**Comment by disksing** [comment]
LGTM

**Comment by morefreeze** [comment]
@siddontang PTAL

**Comment by siddontang** [comment]
LGTM

**Pull request merged**